pkg/services/ngalert/provisioning/alert_rules.go

```diff
@@ -212,10 +308,61 @@ func (service *AlertRuleService) DeleteAlertRule(ctx context.Context, orgID int6
 		return fmt.Errorf("cannot delete with provided provenance '%s', needs '%s'", provenance, storedProvenance)
 	}
 	return service.xact.InTransaction(ctx, func(ctx context.Context) error {
-		err := service.ruleStore.DeleteAlertRulesByUID(ctx, orgID, ruleUID)
-		if err != nil {
-			return err
-		}
-		return service.provenanceStore.DeleteProvenance(ctx, rule, rule.OrgID)
+		return service.deleteRules(ctx, orgID, rule)
+	})
+}
+
+// checkLimitsTransactionCtx checks whether the current transaction (as identified by the ctx) breaches configured alert rule limits.
+func (service *AlertRuleService) checkLimitsTransactionCtx(ctx context.Context, orgID, userID int64) error {
+	limitReached, err := service.quotas.CheckQuotaReached(ctx, "alert_rule", &quota.ScopeParameters{
+		OrgID:  orgID,
+		UserID: userID,
 	})
+	if err != nil {
+		return fmt.Errorf("failed to check alert rule quota: %w", err)
+	}
+	if limitReached {
+		return models.ErrQuotaReached
+	}
+	return nil
+}
+
+// deleteRules deletes a set of target rules and associated data, while checking for database consistency.
+func (service *AlertRuleService) deleteRules(ctx context.Context, orgID int64, targets ...*models.AlertRule) error {
+	uids := make([]string, 0, len(targets))
+	for _, tgt := range targets {
+		if tgt != nil {
+			uids = append(uids, tgt.UID)
+		}
+	}
+	if err := service.ruleStore.DeleteAlertRulesByUID(ctx, orgID, uids...); err != nil {
+		return err
+	}
+	for _, uid := range uids {
+		if err := service.provenanceStore.DeleteProvenance(ctx, &models.AlertRule{UID: uid}, orgID); err != nil {
+			// We failed to clean up the record, but this doesn't break things. Log it and move on.
+			service.log.Warn("failed to delete provenance record for rule: %w", err)
+		}
+	}
+	return nil
+}
+
+// syncRuleGroupFields synchronizes calculated fields across multiple rules in a group.
+func syncGroupRuleFields(group *definitions.AlertRuleGroup, orgID int64) *definitions.AlertRuleGroup {
+	for i := range group.Rules {
+		group.Rules[i].For = (time.Duration(group.Interval) * time.Second)
+		group.Rules[i].IntervalSeconds = group.Interval
+		group.Rules[i].RuleGroup = group.Title
+		group.Rules[i].NamespaceUID = group.FolderUID
+		group.Rules[i].OrgID = orgID
+	}
+	return group
+}
+
+func dropNilAlertRules(ptrs []*models.AlertRule) []models.AlertRule {
```

Comment:
Is this not more of a transformation than a drop? I was confused by the naming at first.

Reply:
Renamed to `withoutNilAlertRules` so it's a little more fluent when calling as a transformation.

Reply:
I'm not sure we are talking about the same thing here. `append` will also append nil values? So it does not remove any nil alert rules?

Reply:
I would expect something like that in the beginning of the `for` loop 

```golang
if ptr == nil {
  continue
}
```

Reply:
Yep, totally forgot this :laughing: fixed!